nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java

```diff
@@ -253,32 +253,46 @@ private void reportInvalidInstantiationError(
   private void reportInvalidAssignmentInstantiationError(
       Tree tree, Type lhsType, Type rhsType, VisitorState state) {
     ErrorBuilder errorBuilder = analysis.getErrorBuilder();
+    String msg = errorMessageForIncompatibleTypesAtPseudoAssignment(lhsType, rhsType, state);
     ErrorMessage errorMessage =
-        new ErrorMessage(
-            ErrorMessage.MessageTypes.ASSIGN_GENERIC_NULLABLE,
-            String.format(
-                "Cannot assign from type "
-                    + prettyTypeForError(rhsType, state)
-                    + " to type "
-                    + prettyTypeForError(lhsType, state)
-                    + " due to mismatched nullability of type parameters"));
+        new ErrorMessage(ErrorMessage.MessageTypes.ASSIGN_GENERIC_NULLABLE, msg);
     state.reportMatch(
         errorBuilder.createErrorDescription(
             errorMessage, analysis.buildDescription(tree), state, null));
   }
 
+  private String errorMessageForIncompatibleTypesAtPseudoAssignment(
+      Type lhsType, Type rhsType, VisitorState state) {
+    String prettyRhsType = prettyTypeForError(rhsType, state);
+    String result =
+        String.format(
+            "incompatible types: %s cannot be converted to %s",
+            prettyRhsType, prettyTypeForError(lhsType, state));
+    if (!ASTHelpers.isSameType(lhsType, rhsType, state)
+        && lhsType.getKind() == TypeKind.DECLARED
+        && rhsType.getKind() == TypeKind.DECLARED) {
+      Symbol.TypeSymbol lhsSym = lhsType.asElement();
+      if (lhsSym instanceof Symbol.ClassSymbol) {
+        Type asSuper =
+            TypeSubstitutionUtils.asSuper(
+                state.getTypes(), rhsType, (Symbol.ClassSymbol) lhsSym, config);
+        if (asSuper != null) {
+          result +=
+              String.format(
+                  " (%s is a subtype of %s)", prettyRhsType, prettyTypeForError(asSuper, state));
```

Comment:
How does this work with deeper inheritance hierarchies? Will going to the direct supertype always be enough information?

Reply:
Yes, it's possible a developer would have to look at the full inheritance path to figure out what is going wrong.  I don't think that's too common of a case.  In most cases I've seen it's just one level.  (This is the case for the error currently being reported on Spring.)  I think in the future we could look at printing the full inheritance path if we see a need for it.